docs: webhook scaling (#1709)
actions-runner-controller/actions-runner-controller: README.md

* Metadata (read)

**Organization Permissions**

* Self-hosted runners (read / write)

_Note: All API routes mapped to their permissions can be found [here](https://docs.github.com/en/rest/reference/permissions-required-for-github-apps) if you wish to review_

**Subscribe to events**

At this point you have a choice of configuring a webhook, a webhook is needed if you are going to use [webhook driven scaling](#webhook-driven-scaling). The webhook can be configured centrally in the GitHub app itself or separately. In either case you need to subscribe to the same events:

> * Check run
> * Workflow job

Reviewer: ```suggestion
* Workflow job
* Check run
```

I think we'd generally recommend `Workflow job` first so that users won't try `check run` and experience various gotchas of it.

Or, perhaps we'd better remove the reference to `check run` at all, assuming all the GHES in the world has already migrated to versions that supports `workflow_job` events? 😁

Under the hood, `RunnerSet` relies on Kubernetes's `StatefulSet` and Mutating Webhook. A `statefulset` is used to create a number of pods that has stable names and dynamically provisioned persistent volumes, so that each `statefulset-managed` pod gets the same persistent volume even after restarting. A mutating webhook is used to dynamically inject a runner's "registration token" which is used to call GitHub's "Create Runner" API.

**Limitations**

* For autoscaling the `RunnerSet` kind only supports pull driven scaling or the `workflow_job` event for webhook driven scaling.

> 

Reviewer: Ah, good catch! Yes, this limitation is obsolete now.

Webhooks are processed by a separate webhook server. The webhook server receives `workflow_job` webhook events and scales RunnerDeployments / RunnerSets by updating HRAs configured for the webhook trigger. Below is an example set-up where a HRA has been configured to scale a `RunnerDeployment` from a `workflow_job` event:

```yaml
apiVersion: actions.summerwind.dev/v1alpha1
kind: RunnerDeployment
metadata:
  name: example-runners
spec:
  template:
    spec:
      repository: example/myrepo
---
apiVersion: actions.summerwind.dev/v1alpha1
kind: HorizontalRunnerAutoscaler
metadata:
  name: example-runners
spec:
  minReplicas: 1
  maxReplicas: 10
  scaleTargetRef:
    kind: RunnerDeployment
    # # In case the scale target is RunnerSet:
    # kind: RunnerSet
    name: example-runners
  scaleUpTriggers:
    - githubEvent:
        workflowJob: {}
      duration: "30m"
```

The lifecycle of a runner provisioned from a webhook is different to a runner provisioned from the pull based scaling method:

1. GitHub sends a `workflow_job` event to ARC with `status=queued`

> 2. ARC finds HRAs with a `workflow_job` webhook scale trigger that and back a RunnerDeployment / RunnerSet with matching runner labels

Reviewer: ```suggestion
2. ARC finds an HRA with a `workflow_job` webhook scale trigger that backs a RunnerDeployment / RunnerSet with matching runner labels
```

Nit but perhaps we should be more explicit about that a `workflow_job` will ever result in scaling only one RunnerDeployment / RunnerSet, by using a singular noun?

> 2. ARC finds HRAs with a `workflow_job` webhook scale trigger that and back a RunnerDeployment / RunnerSet with matching runner labels
> 3. The matched HRAs adds a unit to its `capacityReservations` list

Reviewer: ```suggestion
2. ARC finds an HRA with a `workflow_job` webhook scale trigger that backs a RunnerDeployment / RunnerSet with matching runner labels
3. The matched HRA adds a unit to its `capacityReservations` list
```

Nit but perhaps we should be more explicit about that a `workflow_job` will ever result in scaling only one RunnerDeployment / RunnerSet, by using a singular noun?

> 4. ARC adds a replica and sets the EffectiveTime of that replica to current + `HRA.spec.scaleUpTriggers[].duration`

Reviewer: ```suggestion
4. ARC adds a replica and sets the EffectiveTime of the whole RunnerDeployment to the current time
```

EffectiveTime is used to ensure that the replicas are added only once per each the RunnerDeployment controller detected the changes in the desired `replicas`. Probably this is a bit more correct explanation

At this point there are a few things that can happen, either the job gets allocated to the runner or the runner is left dangling due to it not being used, if the runner gets assigned the job that triggered the scale up the lifecycle looks like this:

1. The new runner gets allocated the job and processes it
2. Upon the job ending GitHub sends another `workflow_job` event to ARC but with `status=completed`

> 3. The HRA removes a capacity reservation from its `capacityReservations` and picks a runner to terminate ensuring it isn't busy via the GitHub API beforehand

Reviewer: ```suggestion
3. The HRA removes the oldest capacity reservation from its `capacityReservations` and picks a runner to terminate ensuring it isn't busy via the GitHub API beforehand
```

so that it won't end up in scaling in/removing unnecessarily more runners.

> If the job is cancelled before it is allocated to a runner or the runner is never used due to other runners matching needed runner group and required runner labels are allocated the job then the lifecycle looks like this:

Reviewer: ```suggestion
If the job is cancelled before it is allocated to a runner then the lifecycle looks like this:

1. Upon the job cancellation GitHub sends another `workflow_job` event to ARC but with `status=cancelled`
2. The HRA removes the oldest capacity reservation from its `capacityReservations` and picks a runner to terminate ensuring it isn't busy via the GitHub API beforehand

If runner is never used due to other runners matching needed runner group and required runner labels are allocated the job then the lifecycle looks like this:

1. The scale trigger duration specified via `HRA.spec.scaleUpTriggers[].duration` elapses
2. The HRA thinks the capacity reservation is expired, removes it from HRA's `capacityReservations` and terminates the expired runner ensuring it isn't busy via the GitHub API beforehand
```

Reviewer: Perhaps this might better reflect the implementation as a job cancellation can trigger an immediate scale down

1. The HRA removes a capacity reservation from its `capacityReservations` and terminates the expired runner ensuring it isn't busy via the GitHub API beforehand

Your `HRA.spec.scaleUpTriggers[].duration` value should be set long enough to account for 2 things:

> 1. the potential amount of time it could take for a pod to become `Running` e.g. you need to scale horizontally because there isn't a node avaliable 
> 2. the amount of time it takes for GitHub to allocate a job to that runner

Reviewer: ```suggestion
1. the potential amount of time it could take for a pod to become `Running` e.g. you need to scale horizontally because there isn't a node avaliable 
2. the amount of time it takes for GitHub to allocate a job to that runner
3. the amount of time it takes for the runner to notice the allocated job and starts running it
```

Reviewer: 3 is also important because, without it, ARC may remove the runner before the runner starts running the job and becomes busy.

1. GitHub sends a `workflow_job` event to ARC with `status=queued`
2. ARC finds HRAs with a `workflow_job` webhook scale trigger that and back a RunnerDeployment / RunnerSet with matching runner labels
3. The matched HRAs adds a unit to its `capacityReservations` list
4. ARC adds a replica and sets the EffectiveTime of that replica to current + `HRA.spec.scaleUpTriggers[].duration`

At this point there are a few things that can happen, either the job gets allocated to the runner or the runner is left dangling due to it not being used, if the runner gets assigned the job that triggered the scale up the lifecycle looks like this:

1. The new runner gets allocated the job and processes it
2. Upon the job ending GitHub sends another `workflow_job` event to ARC but with `status=completed`
3. The HRA removes a capacity reservation from its `capacityReservations` and picks a runner to terminate ensuring it isn't busy via the GitHub API beforehand

If the job is cancelled before it is allocated to a runner or the runner is never used due to other runners matching needed runner group and required runner labels are allocated the job then the lifecycle looks like this:

> 1. The new runner is retained until its `EffectiveTime` passes
> 3. The HRA removes a capacity reservation from its `capacityReservations` and terminates the expired runner ensuring it isn't busy via the GitHub API beforehand

Reviewer: ```suggestion
1. The HRA removes a capacity reservation from its `capacityReservations` and terminates the expired runner ensuring it isn't busy via the GitHub API beforehand
```

`EffectiveTime` is used to prevent creating a redundant runner, in case a runner pod has successfully finished running a job and completed with exist status 0. So I'd say it isn't involved in this scenario.

`EffectiveTime` works as follows:

There is a race between the timing that the runner pod finishes running and the timing that GitHub sends a `status=completed` `workflow_job` event.

When the former happened before the latter(which is almost 100% cases), ARC used to end up with recreating a runner pod because RunnerDeployment specs > available runner pod replicas.

With EffectiveTime and saving the EffectiveTime that ARC saw when it ran the last reconciliation on the RunnerDeployment, ARC can prevent this pod recreation because it now knows that it already created all the desired replicas after seeing that EffectiveTime.
